datalad_metalad/add.py

```diff
@@ -0,0 +1,449 @@
+
+      Add metadata stored in the file "metadata-123.json" to the
+      metadata model instance in the current directory and overwrite
+      the "dataset_id" value stored in "metadata-123.json".
+
+      $ datalad meta-add --metadata-store /metadata/dataset_0 metadata-123.json \
+       '{"dataset_id": "00010203-1011-2021-3031-404142434445}'
+
+      Add metadata read from standard input to the metadata model
+      instance in the current directory.
+
+      $ datalad meta-add -
+
+      All arguments can be pre-fixed by '@', in which case the pre-fixed
+      argument is interpreted as a file-name and the argument value is read
+      from the file. For example, add metadata stored in the file
+      "metadata-123.json" to the metadata model instance in the current
+      directory and overwrite metadata values with the values stored in
+      stored in "extra-info.json".
+
+      $ datalad meta-add --metadata-store /metadata/dataset_0 \
+       metadata-123.json @extra-info.json
+
+    """
+
+    required_keys = (
+        "extractor_name",
+        "extractor_version",
+        "extraction_parameter",
+        "extraction_time",
+        "agent_name",
+        "agent_email",
+        "dataset_id",
+        "dataset_version",
+        "extracted_metadata")
+
+    optional_keys = (
+        "intra_dataset_path",)
+
+    required_additional_keys = (
+        "root_dataset_id",
+        "root_dataset_version",
+        "inter_dataset_path")
+
+    required_keys_lines = "\n".join(map(repr, required_keys))
+    required_additional_keys_lines = "\n".join(
+        map(repr, required_additional_keys))
+
+    _params_ = dict(
+            metadata=Parameter(
+            args=("metadata",),
+            metavar="METADATA",
+            doc=f"""Path of a file that contains the metadata that
+            should be added to the metadata model instance (the
+            metadata must be provided as a JSON-serialized metadata
+            dictionary).
+            
+            If the path is "-", metadata is read from standard input.
+            
+            The dictionary must contain the following keys:
+            
+            {required_keys_lines}
+            
+            If the metadata is associated with a file, the following key
+            indicates the file path:
+            
+            'intra_dataset_path'
+            
+            It may in addition contain either all or none of the
+            following keys (they are used to add metadata element
+            as a sub-dataset element, i.e. perform aggregation):
+            
+            {required_additional_keys_lines}            
+            """,
+            constraints=EnsureStr() | EnsureNone()),
+        metadata_store=Parameter(
+            args=("-s", "--metadata-store"),
+            doc="""Directory in which the metadata model instance is
+            stored. If no directory name is provided, the current working
+            directory is used.""",
```

Comment:
> Directory in which the metadata model instance is store

This must be the path to a Git repository, right?

Reply:
Yes (and no). Since we currently only use git-based model instances, it must be a path to a Git repository. (If we use other backends, it must be a "path" to the backend, could also be something like "odbc://localhost:2289/database1" etc.)

Reply:
Understood. The parameter description should then reflect this generality, I think.

> Metadata store to add to. At present this can only be the path to a local Git repository.

Reply:
Created #97